File: backend-java/src/main/java/kz/savefood/util/Geo.java
package kz.savefood.util;

/** Geodesy helpers, ported from utils.py {@code haversine} — the single implementation. */
public final class Geo {
    private Geo() {
    }

    private static final double EARTH_RADIUS_METERS = 6_371_000;

    /** Great-circle distance in metres. */
    public static double haversineMeters(double lat1, double lon1, double lat2, double lon2) {
        double phi1 = Math.toRadians(lat1);
        double phi2 = Math.toRadians(lat2);
        double dphi = Math.toRadians(lat2 - lat1);
        double dlambda = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dphi / 2) * Math.sin(dphi / 2)
            + Math.cos(phi1) * Math.cos(phi2) * Math.sin(dlambda / 2) * Math.sin(dlambda / 2);
        return 2 * EARTH_RADIUS_METERS * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    }
}

File: savefood/src/utils/geo.js
/** Great-circle distance in metres. Mirrors Geo.haversineMeters on the server. */
export const haversineMeters = (lat1, lon1, lat2, lon2) => {
  const R = 6371000;
  const p1 = (lat1 * Math.PI) / 180;
  const p2 = (lat2 * Math.PI) / 180;
  const dp = ((lat2 - lat1) * Math.PI) / 180;
  const dl = ((lon2 - lon1) * Math.PI) / 180;
  const a = Math.sin(dp / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
};
export const buildNavigatorUrls = (stops) => {
  const list = (Array.isArray(stops) ? stops : [stops]).filter(
    p => Number.isFinite(p?.lat) && Number.isFinite(p?.lon),
  );
  if (list.length === 0) return null;
  const waypoints = list.map(p => `${p.lat},${p.lon}`).join('~');
  const { lat, lon } = list[0];
  return {
    app: `yandexnavi://build_route_on_map?lat_to=${lat}&lon_to=${lon}`,
    web: `https://yandex.ru/maps/?rtext=~${waypoints}&rtt=auto`,
  };
};
